python/lsst/ctrl/mpexec/preExecInit.py

```diff
@@ -125,19 +124,55 @@ def initializeDatasetTypes(self, graph, registerDatasetTypes=False):
         datasetTypes = PipelineDatasetTypes.fromPipeline(
             pipeline, registry=self.butler.registry, include_configs=True, include_packages=True
         )
-        for datasetType in itertools.chain(
-            datasetTypes.initIntermediates,
-            datasetTypes.initOutputs,
-            datasetTypes.intermediates,
-            datasetTypes.outputs,
+
+        for datasetTypes, is_input, is_output in (
+            (datasetTypes.initIntermediates, True, True),
+            (datasetTypes.initOutputs, False, True),
+            (datasetTypes.intermediates, True, True),
+            (datasetTypes.outputs, False, True),
         ):
+            self._register_dataset_types(registerDatasetTypes, datasetTypes, is_input, is_output)
+
+    def _register_dataset_types(self, registerDatasetTypes, datasetTypes, is_input, is_output):
+        def _check_compatibility(datasetType, expected, is_input, is_output) -> bool:
+            is_compatible = False
+            if is_input:
+                # This DatasetType must be compatible on get.
+                is_compatible = datasetType.is_compatible_with(expected)
+            if is_output:
+                # Has to be able to be converted to expect type on put.
+                put_is_compatible = expected.is_compatible_with(datasetType)
+                if is_input:
+                    is_compatible = put_is_compatible and is_compatible
+                else:
+                    is_compatible = put_is_compatible
```

Comment:
This looks more complicated than it should be. Maybe set `is_compatible` to `True` initially? Or is there a case when both `is_input` and `is_output` are false?

Reply:
Cleaned it up.